### DISCUSSION TITLE
Handling `up.CannotMatch` globally

### DISCUSSION THREAD

**Discussion by matteocontrini** [unpoly/unpoly#519] [Q&A]
Hello,

is there a way to catch the `up.CannotMatch` error globally, in order to show an error to the user?

I've tried the following but the `onError` callback never gets called. Am I doing something wrong?

```
up.on('up:fragment:loaded', (event: any) => {
    event.renderOptions.onError = (error: any) => {
        if (error instanceof up.CannotMatch) {
            // show error
        }
    };
});
```

Thanks.

**Comment by adam12** [comment]
I'm not sure there is a way to do this, or what your specific error is, but you could try just doing your own render here perhaps. What kind of scenarios are you expecting where you'd not be able to match a selector?

```javascript
up.on('up:fragment:loaded', (event) => {
  event.preventDefault()

  event.renderOptions.onError = (error) => {
    if (error instanceof up.CannotMatch) {
      // show error
    }
  }

  up.render(event.renderOptions)
})
```

**Reply by matteocontrini** [reply]
Thanks for your reply. I've tried your code but it seems to cause an infinite loop and the browser quickly freezes.

> What kind of scenarios are you expecting where you'd not be able to match a selector?

I have a few overlays where I define the Unpoly main element with `up-main="overlay"`.

When a request triggered by an overlay returns an error page, I see the following error:

`Uncaught (in promise) up.CannotMatch: Could not find common target in current page and response (tried selectors [[up-main~=drawer], [up-main~=overlay], [up-main=''], main, [up-main="overlay"]] in layer "origin")`

I understand that this is because the error response doesn't contain a matching `up-main="overlay"` (which I'm unable to ensure it's always present on all kinds of error pages).

I've tried the following:

1) `up-fail-layer="root"`, which fixes the issue but then I can't make sure validation errors are shown within the overlay and not in the root

2) catching `up.CannotMatch` globally, so that I can at least show an error to the user

Am I missing something? Should I change something in the way I'm doing things here?

Thank you 😃

**Comment by matteocontrini** [comment]
I've found this solution:

```js
up.on('up:fragment:loaded', function (event) {
    const isFailed = up.network.config.fail(event.renderOptions.response);
    if (isFailed && event.response.status != 422) {
        // Force the fail layer or show an error alert
        event.renderOptions.failLayer = 'root';
    }
});
```

It acts before `up.CannotMatch` by "hijacking" the fail layer, only if the status code is not 422. It seems to work, but there might be some situations I haven't considered. And it probably needs adjustments to work well with `up-fail` attributes.

I still think there should be a built-in way to handle these situations though, they seem common enough to me.

**Reply by adam12** [reply]
Nice solution.
